inference-engine/src/mkldnn_plugin/nodes/extract_image_patches.cpp

```diff
@@ -1,248 +1,485 @@
+            default:
+                assert(!"unsupported instruction set in custom_uni_vgatherdps");
+        }
+    }
 
-        const int64_t IC = inDims[CHANNEL];
-        const int64_t IH = inDims[inDimsSize - 2];
-        const int64_t IW = inDims[inDimsSize - 1];
+    inline void gather_src2vmm(const Vmm &vmm_arg, reg64_t &mem_base) {
+        switch (jpp.dtype_size) {
+            case 4: custom_uni_vgatherdps(vmm, mem_base, vmm_gather_index, vmm_gather_mask); break;
+            case 2:
+            case 1: emulate_gather(vmm_arg, mem_base); break;
+            default:
+                assert(!"unknown dtype size for gather");
+        }
+        add(mem_base, jpp.SW * jpp.dtype_size * jpp.block_size);
+    }
 
-        const auto& outDims = outputs[0]->getTensorDesc().getDims();
-        const size_t outDimsSize = outDims.size();
+    inline void emulate_gather(const Xbyak::Xmm &xmm_arg, reg64_t &mem_base, int xmm_offset = 0) {
+        const int xmm_size = 16; // bytes
+        const int xmm_block_size = xmm_size / jpp.dtype_size;
+        const int offset = xmm_offset * jpp.SW * jpp.dtype_size * xmm_block_size;
+        for (int i = 0; i < xmm_block_size; i++) {
+            Xbyak::Address addr = ptr[mem_base + i * jpp.SW * jpp.dtype_size + offset];
+            switch (jpp.dtype_size) {
+                case 4: uni_vpinsrd(xmm_arg, xmm_arg, addr, i); break;
+                case 2: uni_vpinsrw(xmm_arg, xmm_arg, addr, i); break;
+                case 1: uni_vpinsrb(xmm_arg, xmm_arg, addr, i); break;
+                default:
+                    assert(!"unknown dtype size");
+            }
+        }
+    }
+    inline void emulate_gather(const Xbyak::Ymm &ymm_arg, reg64_t &mem_base) {
+        Xbyak::Xmm low_xmm = Xbyak::Xmm(ymm_arg.getIdx());
+        emulate_gather(low_xmm, mem_base, 0);
+        emulate_gather(xmm_aux, mem_base, 1);
+        vinserti128(ymm_arg, ymm_arg, xmm_aux, 1);
+    }
 
-        const int64_t OB = outDims[BATCH];
-        const int64_t OC = outDims[CHANNEL];
-        const int64_t OH = outDims[outDimsSize - 2];
-        const int64_t OW = outDims[outDimsSize - 1];
+    inline void emulate_gather(const Xbyak::Zmm &zmm_arg, reg64_t &mem_base) {
+        Xbyak::Xmm low_xmm = Xbyak::Xmm(zmm_arg.getIdx());
+        emulate_gather(low_xmm, mem_base, 0);
+        for (int i = 1; i < 4; i++) {
+            emulate_gather(xmm_aux, mem_base, i);
+            vinserti64x2(zmm_arg, zmm_arg, xmm_aux, i);
+        }
+    }
 
-        const int64_t KH = _ksizes[HIGHT];
-        const int64_t KW = _ksizes[WIDTH];
-        const int64_t SH = _strides[HIGHT];
-        const int64_t SW = _strides[WIDTH];
-        const int64_t RH = _rates[HIGHT];
-        const int64_t RW = _rates[WIDTH];
+    void loop() {
+        Xbyak::Label ih_loop, ih_tail, ih_exit;
+        Xbyak::Label iw_loop, iw_tail, iw_exit;
+        mov(reg_oh_count, reg_h_hi_pad);
+        mov(reg_num_pads, ptr[reg_params + GET_OFF(h_lo_pad)]);
```

Comment:
The reason was to make it clear that reg_num_pads contains h_lo_pad, otherwise it seems a bit confusing why reg_num_pads (which is usually used as an pad_with_zeros argument) is subtracted from reg_oh_count. Anyway, I don't mind moving the h_lo_pad reading to the generator (so I did), but also I left a comment on why reg_num_pads is used differently this time.